### DISCUSSION TITLE
Compute PackageValidationBaselineVersion with MinVer?

### DISCUSSION THREAD

**Discussion by sliekens** [adamralph/minver#1059] [Ideas]
Hi,

I've been successfully using both MinVer (fantastic tool btw) and [.NET Package Validation](https://learn.microsoft.com/en-us/dotnet/fundamentals/apicompat/package-validation/baseline-version-validator) to check for breaking changes when running `dotnet pack`.

``` csproj
  <PropertyGroup>
    <EnablePackageValidation>true</EnablePackageValidation>
    <PackageValidationBaselineVersion>1.0.0</PackageValidationBaselineVersion>
  </PropertyGroup>
  ```
  
I've been manually updating the `PackageValidationBaselineVersion` property after each release, but I just realized I always just want to keep this in sync with the last package version.

Currently, MinVer sets `$(Version)` and `$(PackageVersion)`, but I think I would need something like `$(PreviousPackageVersion)` to automate this.

``` csproj
  <PropertyGroup>
    <EnablePackageValidation>true</EnablePackageValidation>
    <PackageValidationBaselineVersion>$(PreviousPackageVersion)</PackageValidationBaselineVersion>
  </PropertyGroup>
```

Wouldn't it be possible to make this a computed property using MinVer build tasks?

**Comment by adamralph** [comment]
The main problem that computing that property presents is that the console apps (both the one used internally by the MinVer MSBuild task and the minver-cli dotnet tool) currently return a single string value—the version. They have no facility to return structured data consisting of more than a single literal, e.g a version and a previous version, and I'm not sure I want to introduce that complexity.

Another way to do this would be to write a custom target to make MinVer run in the context of the commit prior to the current commit, store the value of `MinVerVersion`, and then allow MinVer to run as it currently does to get the version of the current commit.

Pseudo-code:

```xml
<Target Name="MyTarget" BeforeTargets="MinVer">
  If there is a commit prior to the current commit
    git checkout HEAD^
    set MinVerIgnoreHeight to true
    run the MinVer target
    set MinVerIgnoreHeight to false
    git checkout @{-1}
    store the value of MinVerVersion in MinVerVersionPrevious    
</Target>
```

This assume you are currently _not_ setting `MinVerIgnoreHeight` to `true` for determining the current version of your package. If you are, then this will be more complex. Maybe something like:

```xml
<Target Name="MyTarget" BeforeTargets="MinVer">
  If there is a commit prior to the current commit
    store the current commit sha
    If MinVerIgnoreHeight is true
        run the MinVer target
        git checkout MinVerVersion
    else
       git checkout HEAD^
    store the original value of MinVerIgnoreHeight
    set MinVerIgnoreHeight to true
    run the MinVer target
    set MinVerIgnoreHeight to the original value
    git checkout {original commit sha}
    store the value of MinVerVersion in MinVerVersionPrevious    
</Target>
```

You may also have to do something to cater for the edge case of not having any tag in the commit history.
